plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java

```diff
@@ -20,61 +20,127 @@
 import static java.util.Locale.ENGLISH;
+import static java.util.Objects.requireNonNull;
 import static java.util.UUID.randomUUID;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
 import static java.util.stream.Collectors.joining;
 
-// holds caches and mappings
-// Trino converts the dataset and table names to lower case, while BigQuery is case sensitive
-// the mappings here keep the mappings
 class BigQueryClient
 {
     private final BigQuery bigQuery;
     private final Optional<String> viewMaterializationProject;
     private final Optional<String> viewMaterializationDataset;
-    private final ConcurrentMap<TableId, TableId> tableIds = new ConcurrentHashMap<>();
-    private final ConcurrentMap<DatasetId, DatasetId> datasetIds = new ConcurrentHashMap<>();
+    private final boolean caseInsensitiveNameMatching;
+    private final Cache<String, String> remoteDatasetNames;
+    private final Cache<String, Map<String, String>> remoteTableNames;
 
     BigQueryClient(BigQuery bigQuery, BigQueryConfig config)
     {
         this.bigQuery = bigQuery;
         this.viewMaterializationProject = config.getViewMaterializationProject();
         this.viewMaterializationDataset = config.getViewMaterializationDataset();
+        Duration caseInsensitiveNameMatchingCacheTtl = requireNonNull(config.getCaseInsensitiveNameMatchingCacheTtl(), "caseInsensitiveNameMatchingCacheTtl is null");
+
+        this.caseInsensitiveNameMatching = config.isCaseInsensitiveNameMatching();
+        CacheBuilder<Object, Object> remoteNamesCacheBuilder = CacheBuilder.newBuilder()
+                .expireAfterWrite(caseInsensitiveNameMatchingCacheTtl.toMillis(), MILLISECONDS);
+        this.remoteDatasetNames = remoteNamesCacheBuilder.build();
+        this.remoteTableNames = remoteNamesCacheBuilder.build();
+    }
+
+    String toRemoteDatasetName(String projectId, String datasetName)
+    {
+        requireNonNull(projectId, "projectId is null");
+        requireNonNull(datasetName, "datasetName is null");
+        verify(CharMatcher.forPredicate(Character::isUpperCase).matchesNoneOf(datasetName), "Expected schema name from internal metadata to be lowercase: %s", datasetName);
+        if (caseInsensitiveNameMatching) {
+            try {
+                @Nullable String remoteDatasetName = remoteDatasetNames.getIfPresent(datasetName);
+                if (remoteDatasetName == null) {
+                    // this might have been just created, reload cache
+                    Map<String, String> mapping = stream(listDatasets(projectId))
+                            // will throw on collisions to avoid ambiguity
+                            .collect(toImmutableMap(dataset -> dataset.getDatasetId().getDataset().toLowerCase(ENGLISH), dataset -> dataset.getDatasetId().getDataset()));
+                    remoteDatasetNames.putAll(mapping);
+                }
+                remoteDatasetName = remoteDatasetNames.getIfPresent(datasetName);
```

Comment:
I think you should pick this from `mapping`. Wouldn't cache be already empty here if ttl is very short?